compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt

```diff
@@ -321,10 +317,17 @@ class MethodInliner(
                     } else {
                         super.visitMethodInsn(opcode, owner, name, desc, itf)
                     }
-                } else if ((!inliningContext.isInliningLambda || isDefaultLambdaWithReification(inliningContext.lambdaInfo!!)) &&
-                    ReifiedTypeInliner.isNeedClassReificationMarker(MethodInsnNode(opcode, owner, name, desc, false))
-                ) {
-                    //we shouldn't process here content of inlining lambda it should be reified at external level except default lambdas
+                } else if (ReifiedTypeInliner.isNeedClassReificationMarker(MethodInsnNode(opcode, owner, name, desc, false))) {
+                    // Consider this arrangement:
+                    //     inline fun <reified T> f(x: () -> Unit = { /* uses `T` in a local class */ }) = x()
+                    //     inline fun <reified V> g() = f<...> { /* uses `V` in a local class */ }
```

Comment:
@pyos Could you add test?
Could you also  check (and add test if not) do we have test when non reified parameter is passed intead of `V`?

Reply:
There are already tests that check reification of default lambdas.

The `V` part is a bit more funky: AnonymousObjectTransformer indiscriminately applies reification to all anonymous objects, including those in lambdas. This is somewhat OK in the IR backend, since it shouldn't touch those objects at all, but in the old backend this caused weird effects when the type parameters' names match. I've added a fix and a test for this that should also make this small part of MethodInliner a bit clearer.